### ISSUE TITLE
Provide design for inheritance from Cadl

### ISSUE THREAD

**Issue by annelo-msft** [Azure/autorest.csharp#2484]
The August MVP target service has instances of polymorphic models with discriminators:
- https://github.com/microsoft/cadl/blob/main/docs/tutorial.md#extends
- https://github.com/microsoft/cadl/blob/main/docs/tutorial.md#polymorphism-with-discriminators

Provide a design for how to generate these in a DPG client Convenience API.

Polymorphic model generation is NOT required for the August MVP, but including this label as a stretch goal.

**Comment by annelo-msft** [comment]
We will be consistent with the current HLC generator at the moment.

We will not address a number of prior test cases that were specific to swagger:
- [ClassThatInheritsFromBaseClassAndSomePropertiesWithBaseClassOverride](https://github.com/Azure/autorest.csharp/blob/feature/v3/test/TestProjects/Inheritance/Inheritance.json#L82) - identical to ClassThatInheritsFromBaseClassAndSomeProperties
- [ClassThatAlsoDefinesBaseClassProperty](https://github.com/Azure/autorest.csharp/blob/feature/v3/test/TestProjects/Inheritance/Inheritance.json#L131) - not used in any tests?
- [ClassThatInheritsFromBaseClassAndSomeProperties](https://github.com/Azure/autorest.csharp/blob/feature/v3/test/TestProjects/Inheritance/Inheritance.json#L71) - not interesting in Cadl because the spread operator handles the SomeProperties case now
- [ClassThatAlsoDefinesBaseClassProperty](https://github.com/Azure/autorest.csharp/blob/feature/v3/test/TestProjects/Inheritance/Inheritance.json#L138) - also not needed anymore
- [Enum Discriminator Types](https://github.com/Azure/autorest.csharp/blob/feature/v3/test/TestProjects/Inheritance/Inheritance.json#L160-L183) - enum discriminators are a REST API antipattern because adding new enum values is a breaking change, per https://github.com/microsoft/api-guidelines/issues/332
- [Extensible Enum Discriminator Types](https://github.com/Azure/autorest.csharp/blob/feature/v3/test/TestProjects/Inheritance/Inheritance.json#L184-L217) - these are not useful unless we expose the extensible enum in the public API, for which there is not currently a requirement.

**Comment by lirenhe** [comment]
I believe the requirement is clear for supporting inheritance and the eng team should be unblocked for this.

**Comment by chunyu3** [comment]
- polymorphism-with-discriminator: will follow the same design as HLC. ( use base type as input parameter or return type, and Deserialize to actual derived type according to discriminator in base type deserialize met
- extends-without-discrminator: Need to confirm if we need to support this. We may not generate those models. Need to confirm if we need cadl-lint-rule to avoid those definition.

**Comment by lirenhe** [comment]
https://gist.github.com/m-nash/7cd8e251257c5883fe71d46960e6ceec

**Comment by lirenhe** [comment]
Close the design
